File: src/3rd_party_slots/rust_metta_bus_client/src/service_bus.rs
use crate::bus_node::BusNode;
use crate::port_pool::PortPool;
use crate::proxy::PatternMatchingQueryProxy;
use crate::types::BoxError;

#[derive(Debug, Default, Clone)]
pub struct ServiceBus {
	service_list: Vec<String>,
	bus_node: BusNode,
	next_request_serial: u64,
}

impl ServiceBus {
	pub fn new(
		host_id: String, known_peer: String, commands: Vec<String>, port_lower: u32,
		port_upper: u32,
	) -> Result<Self, BoxError> {
		let bus_node = BusNode::new(host_id, commands.clone(), known_peer)?;
		PortPool::initialize_statics(port_lower, port_upper)?;
		Ok(Self { service_list: commands, bus_node, next_request_serial: 0 })
	}

	pub fn issue_bus_command(
		&mut self, proxy: &mut PatternMatchingQueryProxy,
	) -> Result<(), BoxError> {
		log::debug!(target: "das", "{} is issuing BUS command <{}>", self.bus_node.node_id(), proxy.command);
		proxy.requestor_id = self.bus_node.node_id();
		self.next_request_serial += 1;
		proxy.serial = self.next_request_serial;
		proxy.proxy_port = PortPool::get_port();
		if proxy.proxy_port == 0 {
			panic!("No port is available to start bus command proxy");
		} else {
			proxy.setup_proxy_node(None, None);
			let mut args = Vec::new();
			args.push(proxy.requestor_id.clone());
			args.push(proxy.serial.to_string());
			args.push(proxy.proxy_node.node_id());
			for arg in &proxy.args {
				args.push(arg.clone());
			}
			match self.bus_node.send_bus_command(proxy.command.clone(), args) {
				Ok(_) => (),
				Err(_) => {
					let mut answer_flow_finished = proxy.answer_flow_finished.lock().unwrap();
					*answer_flow_finished = true;
				},
			}
		}
		Ok(())
	}

	// TODO: Avoiding warnings...
	pub fn service(&self) -> Vec<String> {
		self.service_list.clone()
	}
}

File: 3rd_party_slots/python_client/hyperon_das/service_bus/service_bus.py
import time
import threading
from hyperon_das.service_bus.port_pool import PortPool
from hyperon_das.distributed_algorithm_node.bus_node import BusCommand, BusNode
from hyperon_das.service_bus.proxy import BusCommandProxy
from hyperon_das.logger import log


class ServiceBus:
    def __init__(self, host_id: str, known_peer: str, commands: list[str], port_lower: int, port_upper: int):
        self.service_list: list[str] = []
        self.next_request_serial: int = 1
        self.bus_node = BusNode(host_id, commands, known_peer)
        PortPool.initialize_statics(port_lower, port_upper)

    def issue_bus_command(self, proxy: 'BusCommandProxy') -> None:
        log.info(f"{self.bus_node.node_id} is issuing BUS command <{proxy.command}>")
        proxy.requestor_id = self.bus_node.node_id
        proxy.serial = self.next_request_serial
        self.next_request_serial += 1
        proxy.proxy_port = PortPool.get_port()

        if proxy.proxy_port == 0:
            raise RuntimeError("No port is available to start bus command proxy")
        else:
            proxy.setup_proxy_node()

            args = [
                proxy.requestor_id,
                str(proxy.serial),
                proxy.proxy_node.node_id
            ]

            proxy.pack_command_line_args()

            for arg in proxy.args:
                args.append(arg)

            try:
                self.bus_node.send_bus_command(proxy.command, args)
                time.sleep(1)
            except Exception:
                proxy.abort_flag = False
                log.error(f"Failed to issue BUS command <{proxy.command}> with args {args}")
                raise


class ServiceBusSingletonMeta(type):
    """
    This is a thread-safe implementation of Singleton.
    """

    _instances = {}

    _lock = threading.Lock()

    def __call__(cls, *args, **kwargs):
        with cls._lock:
            if cls not in cls._instances:
                instance = super().__call__(*args, **kwargs)
                cls._instances[cls] = instance
        return cls._instances[cls]


class ServiceBusSingleton(metaclass=ServiceBusSingletonMeta):
    def __init__(self, host_id: str, known_peer: str, port_lower: int, port_upper: int) -> None:
        self.service_bus = ServiceBus(
            host_id,
            known_peer,
            [
                BusCommand.PATTERN_MATCHING_QUERY,
                BusCommand.ATOMDB
            ],
            port_lower,
            port_upper
        )

    def get_instance(self) -> 'ServiceBus':
        return self.service_bus
